### PR TITLE
4996 Enhance `ConfigParser` to support nested new ID

### LINKED ISSUE #4996: `ConfigParser` can't set new nested ID
**Is your feature request related to a problem? Please describe.**
Currently, our `ConfigParser` can't set new nested ID, for example "key1#2#key2" if even "key1" doesn't exist:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L159

Would be nice to support new nested ID.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #4996 .

### Description
This PR enhanced the `ConfigParser` to support nested new ID.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by mingxin-zheng** [comment]
Looks good to me!

File: monai/bundle/config_parser.py [hunk 1]
```diff
@@ -154,9 +154,25 @@ def __setitem__(self, id: Union[str, int], config: Any):
...
+        conf_ = self.get()
+        if recursive:
+            if conf_ is None:
+                conf_ = [] if keys[0] == "0" else {}
+            for i, k in enumerate(keys[:-1]):
+                if isinstance(conf_, dict):
+                    if k not in conf_:
+                        conf_[k] = [] if keys[i + 1] == "0" else {}
+                if isinstance(conf_, list):
+                    if int(k) == len(conf_):
```
Comment (wyli) [inline monai/bundle/config_parser.py:166]: I think this way of appending new elements to a list by index is quite confusing...

Comment (Nic-Ma) [reply monai/bundle/config_parser.py]: Hi @wyli ,

Thanks for your review.
I added docstring to describe the feature, it's useful in the data analysis of Auto3D to predefine the report template: https://github.com/Project-MONAI/MONAI/pull/4765#discussion_r955224503, CC @mingxin-zheng .
And another typical use case is to add a transform or handler when overriding a config file.
Refer to Hydra argument: https://github.com/facebookresearch/hydra/blob/v1.2.0/hydra/_internal/instantiate/_instantiate2.py#L155

Thanks.

Comment (wyli) [reply monai/bundle/config_parser.py]: it doesn't make sense in terms of the python style...
it's not ok to set an array with a new index
```py
array = [0, 1]
array[2] = 2  # IndexError: list assignment index out of range
```
while it's ok to set a dictionary with a new key:
```py
d = {"a": 0}
d['b'] = 1
```

Comment (Nic-Ma) [reply monai/bundle/config_parser.py]: Hi @wyli ,

Maybe I didn't describe the PR clearly.
The usage is not `array[2] = 2`, which is against python style.
Instead, I just added the `recursive` option to the `set()` function, similar to the enhancement on `get()`, we added support to give the default value in the `get()` function.
So the usage will be:
```py
array = ConfigParser([0, 1])
array.set(id="2") = 2

array[3] = 3  # IndexError
```
What do you think?

Thanks.


Thanks.

Comment (Nic-Ma) [reply monai/bundle/config_parser.py]: Simplified to only support nested dicts.

Thanks.

**Pull request merged**